Automattic/jetpack
# Add an image thumbnail to wp-admin post list.

We've created a Composer package in Jetpack to add an image thumbnail to the post-list wp-admin page similar to Calypso. For the thumbnail, we use the "featured image" if one exists, else we look in the post itself for the first media-library hosted image.

## Design

![image](https://user-images.githubusercontent.com/77539/129704120-54dd7244-704e-4677-a708-58a4a11e36c5.png)

#### Does this pull request change what data or activity we track or use?

No

## Testing instructions:
### Setup Environment
1. Checkout this branch
```bash
git checkout update/add-posts-list-featured-image
```
2. Add this dependency to Jetpack plugin
```bash
cd projects/plugins/jetpack && composer require automattic/jetpack-post-list
```
3. Update pnpm
```bash
pnpm install
```
4. Update Jetpack dependencies
```bash
jetpack install -r && jetpack install --all
```
5. Run posts-list package compiling process
```bash
jetpack build packages/post-list
```
6. Run `watch` in case you'd like to edit the code in the client.
```bash
jetpack watch packages/post-list
```
7. In a new terminal run the web server
```
jetpack docker up
```

### Compiling

#### Production

```cli
jetpack build -p packages/post-list
```

#### Development
```cli
jetpack build packages/post-list
```

#### Development watching mode 👀 
```cli
jetpack watch packages/post-list
```

### Unit Tests

To run the package unit tests:

1. go the post-list folder `cd projects/packages/post-list`
2. and run the command `composer test-php`.

If you're running the PHP interpreter that shipped with your Mac you may get the error: `/usr/bin/php declares an invalid value for PHP_VERSION`. [StackOverflow has the solution here](https://stackoverflow.com/questions/68245876/phpunit-does-not-work-when-i-run-the-tests-on-the-laravel-framework-i-get-the-e).

### Load The Post List Package
To load the post list package you need to call it from the ["Must Use Plugins" directory](https://github.com/Automattic/jetpack/blob/master/tools/docker/README.md#must-use-plugins-directory). To do that:

1. Create a file named `0-functions.php` in the `tools/docker/mu-plugins/` directory.
2. In that file add this code: `add_action( 'admin_init', array( '\Automattic\Jetpack\Post_List\Post_List', 'configure' ) );`

### Test Feature
1. Go to http://localhost/wp-admin/edit.php?post-list=true
 Note that post-list=true must be set in the URL.
2. Create a post with a featured image.
3. Create another post without a featured image, but add an image to the post content.
4. You should see those images as thumbnails on the post list page http://localhost/wp-admin/edit.php?post-list=true

![image](https://user-images.githubusercontent.com/77539/130856372-8f66dcb4-324b-4d34-83c2-8c72aede4891.png)
